### ISSUE TITLE
BSC compilation/speed issues

### ISSUE THREAD

**Issue by rnitya** [B-Lang-org/bsc#520]
I have a large design that compiles fine with a mid-2020 version (build 9f4a447). However, there are issues in several of the later versions starting from mid-2021. I have tried a few different ones including the latest version: version 2022.01-38-g2005df70 (build 2005df70).

- A large file with types declarations (structures and enums) doesn't compile at all (memory usage keeps increasing and the process has to be killed eventually beyond 10GB of use). I had to split that into two files to make it compile. No FShow is used for the large enums. 
- The compilation is quite slow, I have a > 2X slowdown from the new compiler compared to the old one.

I understand a part of this may be related to issue #334 but I also wanted to bring up the large files issue.

**Comment by quark17** [comment]
Thanks for reporting this, and sorry that you are encountering it.  I would guess that it's the same issue.  And it's a serious regression that we really need to fix, but haven't had the resources unfortunately.

**Comment by quark17** [comment]
As mentioned on Discussion #414, defining an explicit `Generic` instance for large enum types can be a workaround.

**Comment by quark17** [comment]
Are you able to provide files that demonstrate the two issues (failure to compile and slow to compile)? Or provide more details about what constitutes a large file?

I created a large file using a Perl script like this:
```
#!/usr/bin/perl
open(FH, '>', "Test.bsv") or die $!;
$i = 0;
while ($i < 1000) {
    print FH "typedef enum { Red$i, Green$i, Blue$i } Foo$i deriving (Eq, Bits);\n";
    $i ++;
}
close FH;
```
to generate a file of lines like this:
```
typedef enum { Red0, Green0, Blue0 } Foo0 deriving (Eq, Bits);
typedef enum { Red1, Green1, Blue1 } Foo1 deriving (Eq, Bits);
typedef enum { Red2, Green2, Blue2 } Foo2 deriving (Eq, Bits);
...
```
For 1000 declarations, it only takes 9 seconds to compile.  For 10,000, I do see that it takes 10 minutes to compile (112 seconds to create the symbol table the first time, 90 seconds the next, 240 seconds in typecheck, then 110 in the third symbol table construction).  These are scaling issues that can be looked at.  But I don't know if that's the slow compile issue that you're encountering -- I would guess you have less type declarations, but each one might be larger in size?  And I still don't have an example of running out of memory.

If you're able to provide source files, that'd be easiest to diagnose.

**Comment by rnitya** [comment]
Thanks for the following up. Sorry, I haven't had the time to try the generics solution yet. I just read the other thread. I will try that first and if that doesn't work, I will share the source file that has this issue.

**Comment by rnitya** [comment]
Update: I had already split the really large types file into two, _types_file1_ with most of the types and a 2nd file _types_file2_ just holding a large enum (150-200 elements) and some smaller structures. Since we have moved forward by several versions, I haven't gone back to the single large file. We are happy with this for now.

For this split version, compilation of _types_file2_ used to take around 12 minutes and the overall build, around 40-45 minutes with the recent bsc versions including the latest July 2023 release for Ubuntu 18.04 from here (https://github.com/B-Lang-org/bsc/releases). 

I applied your generics solution from here (https://github.com/B-Lang-org/bsc/discussions/414#discussioncomment-1422278) and it is now down to 1-2 seconds for _types_file2_ and  about 22 minutes for the overall core. This is significantly better and closer to the numbers with a mid-2020 version (build https://github.com/B-Lang-org/bsc/commit/9f4a447f8fafe0eec4200c5495e7df14fb2af88d) that took 17-18 minutes for the overall core build.

Thanks for the help, much appreciated.
